### ISSUE TITLE
net/smtp: Add smt.SendMailWithDialer and make Client.localName public

### ISSUE THREAD

**Issue by g13013** [golang/go#40034]
`smtp.SendMail` actually does a great job, but the fact that it uses the `smtp.Dial` directly make it less flexible both for testing and for certain cases like ours, where we wanted to customize the `Client.localName`

in fact, in certain operations, `Client.localName` must correspond to the FQDN hostname sending the email, this lack of flexibility led us to duplicate the entire SendMail (and tests) code just to address this issue.

I propose that we add a new public method `smtp.SendMailWithDialer` and make the `Client.localName` public, 

```go

type Client struct {
...
	LocalName  string // the name to use in HELO/EHLO
        
}

// note: `smtp.Dial` has the same signature and by default sets `localName: "localhost"`
type Dialer Dial(addr string) (*Client, error)

func SendMailWithDialer(dialer Dialer, addr string, a Auth, from string, to []string, msg []byte) error
```

the change set required is very minimal, `SendMailer` could use `SendMailWithDialer` just by passing `smtp.Dial`, but
would allow much more control both for testing and real uses cases with minimum code, for example:

```go
func MyDialer(addr string) (*Client, error) {
    cl, err := Dial("remoteSmtp:25")
    if err != nil { return nil, err }
    cl.LocalName = "real smtp sender FQDN"
    return cl, nil
}

...
smtp.SendMailWithDialer(MyDialer, addr, auth, from, to, msg)
```

**Comment by g13013** [comment]
Here is the diff required (without the doc change)

```diff
diff --git a/src/net/smtp/smtp.go b/src/net/smtp/smtp.go
index e4e12ae5ee..bd4f8db188 100644
--- a/src/net/smtp/smtp.go
+++ b/src/net/smtp/smtp.go
@@ -46,6 +46,14 @@ type Client struct {
 	helloError error  // the error from the hello
 }
 
+type Dialer func() (*Client, error)
+
+func makeDialer(addr string) func() (*Client, error) {
+	return func() (*Client, error) {
+		return Dial(addr)
+	}
+}
+
 // Dial returns a new Client connected to an SMTP server at addr.
 // The addr must include a port, as in "mail.example.com:smtp".
 func Dial(addr string) (*Client, error) {
@@ -316,7 +324,7 @@ var testHookStartTLS func(*tls.Config) // nil, except for tests
 // attachments (see the mime/multipart package), or other mail
 // functionality. Higher-level packages exist outside of the standard
 // library.
-func SendMail(addr string, a Auth, from string, to []string, msg []byte) error {
+func SendMailWithDialer(dial Dialer, a Auth, from string, to []string, msg []byte) error {
 	if err := validateLine(from); err != nil {
 		return err
 	}
@@ -325,7 +333,7 @@ func SendMail(addr string, a Auth, from string, to []string, msg []byte) error {
 			return err
 		}
 	}
-	c, err := Dial(addr)
+	c, err := dial()
 	if err != nil {
 		return err
 	}
@@ -373,6 +381,10 @@ func SendMail(addr string, a Auth, from string, to []string, msg []byte) error {
 	return c.Quit()
 }
 
+func SendMail(addr string, a Auth, from string, to []string, msg []byte) error {
+	return SendMailWithDialer(makeDialer(addr), a, from, to, msg)
+}
+
 // Extension reports whether an extension is support by the server.
 // The extension name is case-insensitive. If the extension is supported,
 // Extension also returns a string that contains any parameters the
```

**Comment by ianlancetaylor** [comment]
@g13013 Please only send patches using Gerritt or as a GitHub pull request, as described at http://golang.org/doc/contribute.html.  Please don't post patches on the issue tracker.  That avoids any concerns about copyright.  Thanks.

**Comment by g13013** [comment]
@ianlancetaylor oh didn't know about that, sorry !

**Comment by rsc** [comment]
Originally, smtp.SendMail was a tiny little wrapper around the Client type, a convenience.
It has gotten rather large as TLS and various validations were added.

@bradfitz points out maybe it would make sense to add a Client.SendMail method.
Then you can make the Client c however you want and call c.SendMail.
And it avoids a callback in favor of much more direct code.

Thoughts?

**Comment by g13013** [comment]
@rsc interesting, sounds good to me. I think it's even better than the original proposal.

**Comment by g13013** [comment]
Last though, 

If I want to customize the TLS configuration to add a valid CA, a naive would be to override Client.StartTLS, but since the Client.conn & Client.serverName are hidden and initialized only via smtp.Dial, extending would not be possible. An example of this could be found in the [smtp.SendMail](https://github.com/golang/go/blob/master/src/net/smtp/smtp.go#L338) where the hook has been added for testing only and used [here](https://github.com/golang/go/blob/master/src/net/smtp/smtp_test.go#L882).

A solution to this would be to add Client.Dial ? which does pretty much what smtp.Dial and smtp.NewClient do all together, this would make the Client fully extensible and even make the Client instance reusable for further operations, if so, this would make smtp.Dial, smtp.NewClient and smtp.SendMail wrappers only for Client

We could even add Client.DialAndSend for convenience, but this is not required for this issue.

what do you think ?

**Comment by g13013** [comment]
I thought a CL with all possible changes would be better to visualize the needs.

**Comment by rsc** [comment]
@g13013, thanks for sketching the API. Let's leave DialAndSend out, since that can be done as two separate calls.

I don't quite understand the need for c.Dial. A Client is an _already connected_ connection. You get one by calling smtp.Dial or by using net.Dial + smtp.NewClient. It doesn't seem necessary to expand the API to contemplate a not-yet-dialed Client.

I think I followed what you said about wanting to modify the TLS config used by STARTTLS, but I don't see how any of the new API in the example CL enables that flexibility. If we need that flexibility, it seems like we could export the tls config as a field in the Client. Then you use smtp.Dial or net.Dial+smtp.NewClient to get a Client c, set up the TLS config, and call c.SendMail. The new exported field is the only change that seems to be necessary (along with c.SendMail as discussed earlier) to enable custom TLS configs.

Do I have this right?

**Comment by g13013** [comment]
> I think I followed what you said about wanting to modify the TLS config used by STARTTLS, but I don't see how any of the new API in the example CL enables that flexibility

You are absolutely right, I've put all the problems that I see in the Client in one CL, and mislead you with my comment about TLS, it was obviously a poor argumentation.

One note about the change, before moving SendMail to Client, the validation occurs before the Dial, now, the validation of localName and recipients occur always after the Dial, . it could be harmful in the sense that the IP could get very poor [reputation](https://serversmtp.com/ip-score-and-reputation/) in case of an attack, is including an example to enforce best practices sufficient ? Also, it feels odd that the same validation in this case will occur multiple times, before the dial and after from within SendMail and Hello !

**Comment by rsc** [comment]
> Or can we conside keeping the same signature of Client.SendMail, with addr and make it dial for a new connection if addr is not empty

It seems a bit odd to do that - right now all Clients are pre-Dialed. You'd be introducing a new kind of Client, for not much win. The validation happening after the Dial doesn't seem like a huge problem. Connections get lost all the time, and how often does SendMail get called with invalid arguments?

**Comment by rsc** [comment]
I've retitled this to note that the changes are (1) add a TLSConfig \*tls.Config field to smtp.Client, like in http.Server; and (2) add a SendMail method.

**Comment by g13013** [comment]
> and how often does SendMail get called with invalid arguments?
not often I guess, unless hackers involved

> right now all Clients are pre-Dialed. You'd be introducing a new kind of Client
emm, that's a good point


ok then, I think everything have been said, if it's accepted I'll update the CL insha'Allah. thanks @rsc

**Comment by rsc** [comment]
Based on the discussion above, this (add Client.TLSConfig field and Client.SendMail method) seems like a **likely accept**.

**Comment by dcormier** [comment]
How would a new exposed field for `*tls.Config` on `Client` interact with the existing [`(*Client).StartTLS(*tls.Config)`](https://pkg.go.dev/net/smtp?tab=doc#Client.StartTLS) method that allows you to specify the `*tls.Config` to use for `STARTTLS`?

**Comment by rsc** [comment]
It seems like if you call StartTLS manually with a TLS config, that would override the thing in the struct.

**Comment by rsc** [comment]
No change in consensus, so accepted.

**Comment by g13013** [comment]
Sorry for the delay, I had a lot of work to finish these last days.

I've updated the CL with the needed changes, let me know of changes are required.

**Comment by g13013** [comment]
Is there anything new for the CL?
